src/features/device-state/routes/state.ts

```diff
@@ -25,6 +26,137 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
+	release: AnyObject,
+	config: Dictionary<string>,
+): Promise<App> {
+	let composition: AnyObject = {};
+	const services: App['services'] = {};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch {
+			composition = {};
+		}
+	}
+
+	(release.contains__image as AnyObject[]).forEach((ipr) => {
+		// extract the per-image information
+		const image = ipr.image[0];
+
+		const si = serviceInstallFromImage(device, image);
+		if (si == null) {
+			throw new Error(
+				`Could not find service install for device: '${
+					application.uuid
+				}', image: '${image?.id}', service: '${JSON.stringify(
+					image?.is_a_build_of__service,
+				)}', service installs: '${JSON.stringify(device.service_install)}'`,
+			);
+		}
+		const svc = si.service[0];
+
+		const environment: Dictionary<string> = {};
+		varListInsert(ipr.image_environment_variable, environment);
+		varListInsert(application.application_environment_variable, environment);
+		varListInsert(svc.service_environment_variable, environment);
+		varListInsert(device.device_environment_variable, environment);
+		varListInsert(si.device_service_environment_variable, environment);
+
+		const labels: Dictionary<string> = {};
+		[...ipr.image_label, ...svc.service_label].forEach(
+			({ label_name, value }: { label_name: string; value: string }) => {
+				labels[label_name] = value;
+			},
+		);
+
+		_.each(ConfigurationVarsToLabels, (labelName, confName) => {
+			if (confName in config && !(labelName in labels)) {
+				labels[labelName] = config[confName];
+			}
+		});
+
+		const imgRegistry =
+			image.is_stored_at__image_location +
+			(image.content_hash != null ? `@${image.content_hash}` : '');
+
+		services[svc.id] = {
+			imageId: image.id,
+			serviceName: svc.service_name,
+			image: formatImageLocation(imgRegistry),
+			// This needs spoken about...
+			running: true,
+			environment,
+			labels,
+		};
+		// Don't send a null contract as this is a waste
+		// of bandwidth (a null contract is the same as
+		// the lack of a contract field)
+		if (image.contract != null) {
+			services[svc.id].contract = image.contract;
+		}
+
+		if (
+			composition != null &&
+			composition.services != null &&
+			composition.services[svc.service_name] != null
+		) {
+			const compositionService = composition.services[svc.service_name];
+			// We remove the `build` properly explicitly as it's expected to be present
+			// for the builder, but makes no sense for the supervisor to support
+			delete compositionService.build;
+			services[svc.id] = {
+				...compositionService,
+				...services[svc.id],
+			};
+		}
+	});
+
+	return {
+		releaseId: release.id,
+		commit: release.commit,
+		name: application.app_name,
+		services,
+		networks: composition?.networks || {},
+		volumes: composition?.volumes || {},
+		appId: application.id,
```

Comment:
This looks to be new?

Reply:
No, it isn't.  It's just part of the refactoring.